app/controllers/concerns/form_errors.rb

```diff
@@ -0,0 +1,32 @@
+module FormErrors
+  extend ActiveSupport::Concern
+
+  included do
+  rescue_from ActiveRecord::RecordNotUnique, with: :not_unique_response
+  rescue_from URI::InvalidURIError, with: :invalid_uri_response
+  end
+
+  private 
+
+  def not_unique_response(e)
+    response = check_unique_error(e)
+    resource.errors.add(:base, response)
+    logger.warn "Unable to create resource due to '#{e.message}'"
+    render :new, status: :unprocessable_entity
+  end
+
+  def check_unique_error(e)
+    if e.message.include?("canonical_id")
+      "The Canonical ID is already in use for this organization."
+    elsif e.message.include?("source_uri")
+      "The Source URI is already in use for this organization."
+    end
```

Comment:
can you check whether `e` contains attributes that identify the field with the error? that would be more robust than checking the message text. I realize you're jumping into the middle of the code base here, so it may not be possible or worthwhile to effect that change...yet...I'm just wondering if there's a way to do something like `e.fields.key?(:canonical_id)` or similar

Reply:
Mike, I am not seeing any additional information or properties on the error that would allow me to do that. Here is an example of the e object when for the error of identical source_uri.

#<ActiveRecord::RecordNotUnique: PG::UniqueViolation: ERROR:  duplicate key value violates unique constraint "index_resources_on_source_uri_and_organization_id"
DETAIL:  Key (source_uri, organization_id)=(https://cdn.pixabay.com/photo/2023/12/05/01/15/desert-8430551_1280.jpg, 1) already exists.
>

Reply:
yeah you're right - that's a fairly generic wrapper class